Use 1D interpolation build up multidimensional filters

mgiammar opened the pull request:
Pull request seeks to simplify the calculation an construction of multi-dimensional filters by adopting the following workflow

`input` --> `hist_with_lerp` --> `curve_1d` --> `lerp_to_ndim` --> `curve_nd`

- Adds function `utils.bin_1dim_with_lerp` to take multidimensional values with associated coordinates and bins them onto a 1-dimensional grid with linear interpolation.
- Adds function `utils.curve_1dim_to_ndim` which takes in values and frequencies in 1-dimension and places those values onto a multidimensional frequency grid.

mgiammar commented:
Line 190 in whitening.py is throwing unshapeable type list error. Resolve by changing the line to `        image_shape=tuple(real_space_shape),`

mgiammar commented:
@jdickerson95 What would your thoughts be on adding a maximum resolution  argument in terms of inverse pixels or Nyquist to the whitening filter function? Only elements below that resolution are used to calculate the filter, but during interpolation all elements above are set to either `1.0` or `0.0`. Would allow us to quickly swap between Nyquist definition of whitening and how cisTEM implements it out to the corner frequency.

jdickerson95 commented:
Sounds good, could even to a frequency range. We could exclude those values from the max with something like:

jdickerson95 commented:
exclude_indices = (binning_freqs < freq_min) | (binning_freqs > freq_max)
    max_value = whitening_filter_1d[~exclude_indices].max()
    whitening_filter_1d /= max_value

jdickerson95 commented:
And then at the end set to 1. 
    whitening_filter_ndim = torch.where(
        (freq_grid < freq_min) | (freq_grid > freq_max),
        torch.tensor(1.0, device=image_dft.device),
        whitening_filter_ndim,
    )
Could do the normalization in power_spectral_density to have access to binned_freqs

mgiammar commented:
> ```
> exclude_indices = (binning_freqs < freq_min) | (binning_freqs > freq_max)
> max_value = whitening_filter_1d[~exclude_indices].max()
> whitening_filter_1d /= max_value
> ```

Added a version of this to the most recent commit. Will refactor the tests for whitening filter.

mgiammar commented:
Note that I've removed the Gaussian smoothing option from `whitening_filter`. I presume that this could just be done in 1D and simplified a lot (e.g. just some torch conv1d function).

jdickerson95 commented:
Why has the output shape changed from being in real space to Fourier space? I think I prefer giving it in real space.

mgiammar commented:
> Why has the output shape changed from being in real space to Fourier space? I think I prefer giving it in real space.

Which line are you referring to? If it's the following argument in the whitening filter, that's because IMO it is easier to directly pass `output_shape =some_tensor_in_fourier_space.shape` rather than have it on the user keep track of (and possibly incorrectly pass) the real-space shape and potentially RFFT shape of a filter.

```
    output_shape: tuple[int, ...], optional
        The shape of the output filter. If None, then the shape is the same as
        `image_dft`. The default is None. NOTE: This shape is in Fourier space.
```

But you're right to keep with the style of the rest of tt*/torch-* packages this should be passed as the real-space shape.
